### PR TITLE
api: Mark messages as read (markAllAsRead, markStreamAsRead, markTopicAsRead)

### PR DESCRIPTION
**Pull request by sirpengi**
This pr was split off from #361.

Also fixes part of #130

### CODE REVIEW

File: lib/api/route/messages.dart [hunk 1]
```diff
@@ -291,3 +290,148 @@ Future<void> removeReaction(ApiConnection connection, {
...
+// by `updateMessageFlagsForNarrow`) so there are only
+// two versions (FL 153 and FL 154) affected.
+Future<void> markAllAsRead(ApiConnection connection) {
+  return connection.post('markAllAsRead', (_) {}, 'mark_all_as_read', {});
+}
+
+/// https://zulip.com/api/mark-stream-as-read
+///
+/// This binding is deprecated, in FL 155+ use
+/// updateMessageFlagsForNarrow instead.
```
Comment (gnprice) [inline lib/api/route/messages.dart:414]: nit: linkify:
```suggestion
/// [updateMessageFlagsForNarrow] instead.
```

File: test/api/route/messages_test.dart [hunk 2]
```diff
@@ -436,4 +451,196 @@ void main() {
...
+
+  group('markStreamAsRead', () {
+    Future<void> checkMarkStreamAsRead(
+      FakeApiConnection connection, {
+      required int streamId,
+      required Map<String, String> expected,
+    }) async {
+      connection.prepare(json: {});
+      await markStreamAsRead(connection, streamId: streamId);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
```
Comment (gnprice) [inline test/api/route/messages_test.dart:603]:
```suggestion
      check(connection.lastRequest).isA<http.Request>()
```

As I mentioned in our call yesterday, the `isNotNull` is redundant with the subsequent `isA`. The type we're checking for with `isA` already excludes null.

(I think I originally introduced this redundant pattern to this file. But we should stop propagating it.)

**Review by gnprice — COMMENTED** [review]
Thanks! Looks good modulo two small comments (and the completion of #361, which this is stacked upon).

**Comment by sirpengi** [comment]
@gnprice this is ready as well!

**Comment by gnprice** [comment]
Thanks! Looks good; merging.

File: test/api/route/messages_test.dart (hunk 2 above)
Comment (gnprice) [reply test/api/route/messages_test.dart]: → #369 for the existing cases of that pattern.

**Pull request merged**